src/hexkit/providers/akafka/provider/eventsub.py

```diff
@@ -177,76 +210,210 @@ async def construct(
+        topic = event.topic
+        if topic == self._retry_topic:
+            topic = str(event.value.get(ORIGINAL_TOPIC_FIELD, ""))
+            logging.info(
+                "Received previously failed event from topic '%s' for retry.", topic
+            )
+        return topic
+
+    async def _publish_to_dlq(self, *, event: ExtractedEventInfo):
+        """Publish the event to the DLQ topic."""
+        dlq_payload = {**event.payload, ORIGINAL_TOPIC_FIELD: event.topic}
+        logging.debug("About to publish an event to DLQ topic '%s'", self._dlq_topic)
+        await self._publisher.publish(  # type: ignore
+            payload=dlq_payload,
+            type_=event.type_,
+            topic=self._dlq_topic,
+            key=event.key,
         )
+        logging.info("Published event to DLQ topic '%s'", self._dlq_topic)
+
+    async def _retry_event(self, *, event: ExtractedEventInfo, retries_left: int):
+        """Retry the event until the maximum number of retries is reached."""
+        retries_left -= 1
+        try:
+            logging.info(
+                "Retry %i of %i for event of type '%s' on topic '%s' with key '%s'.",
+                self._max_retries - retries_left,
+                self._max_retries,
+                event.type_,
+                event.topic,
+                event.key,
+            )
+            await self._translator.consume(
+                payload=event.payload,
+                type_=event.type_,
+                topic=event.topic,
+                key=event.key,
+            )
+        except Exception as err:
+            if retries_left > 0:
+                await self._retry_event(event=event, retries_left=retries_left)
+            else:
+                raise self.RetriesExhaustedError(
+                    event_type=event.type_, max_retries=self._max_retries
+                ) from err
+
+    async def _handle_consumption(self, *, event: ExtractedEventInfo):
+        """Try to pass the event to the consumer.
+
+        If the event fails:
+        1. Retry until retries are exhausted, if retries are configured.
+        2. Publish the event to the DLQ topic if the DLQ is enabled. Done afterward.
+           or
+        3. Allow failure with unhandled error if DLQ is not configured.
+        """
+        try:
+            await self._translator.consume(
+                payload=event.payload,
+                type_=event.type_,
+                topic=event.topic,
+                key=event.key,
+            )
+        except Exception:
+            logging.warning(
+                "Failed initial attempt to consume event of type '%s' on topic '%s' with key '%s'.",
+                event.type_,
+                event.topic,
+                event.key,
+            )
+            if self._max_retries > 0:
+                # Don't raise RetriesExhaustedError unless retries are actually attempted
```

Comment:
I think we should add some kind of exponential backoff here instead of retrying immediately. This should also be made configurable.

Reply:
Agreed. Added in [fcc9a71](https://github.com/ghga-de/hexkit/pull/123/commits/fcc9a71f40620c5a9fcb80fc17385ff9d2c03208)